File: dotnet/src/LegoCatalog.App/Services/AzureBlobImageStore.cs
using Azure;
using Azure.Identity;
using Azure.Storage.Blobs;
using LegoCatalog.App.Configuration;

namespace LegoCatalog.App.Services;

/// <summary>
/// Reads catalog images from Azure Blob Storage using the workload identity.
/// </summary>
public sealed class AzureBlobImageStore : IImageStore
{
    private readonly BlobContainerClient _container;

    public AzureBlobImageStore(CatalogRuntimeOptions options)
    {
        if (options.BlobServiceEndpoint is null || options.BlobContainerName is null)
        {
            throw new InvalidOperationException(
                "Azure Blob image storage requires CATALOG_BLOB_SERVICE_ENDPOINT and CATALOG_BLOB_CONTAINER.");
        }

        var credential = new DefaultAzureCredential(
            new DefaultAzureCredentialOptions
            {
                ManagedIdentityClientId = options.WorkloadIdentityClientId,
            });
        _container = new BlobServiceClient(
                new Uri(options.BlobServiceEndpoint),
                credential)
            .GetBlobContainerClient(options.BlobContainerName);
    }

    public string GetImageUrl(string fileName) => $"/images/{fileName}";

    public async Task<ReadOnlyMemory<byte>?> ReadAsync(
        string fileName,
        CancellationToken cancellationToken)
    {
        // The canonical-key check is the traversal control, and it has to run here
        // too. ImageSecurityTests only exercises the static validator through
        // LocalImageStore, so a blob store that skipped this would keep the whole
        // suite green while accepting keys the local store rejects.
        if (!LocalImageStore.IsCanonicalImageKey(fileName))
        {
            return null;
        }

        try
        {
            var response = await _container
                .GetBlobClient(fileName)
                .DownloadContentAsync(cancellationToken);
            return response.Value.Content.ToMemory();
        }
        catch (RequestFailedException exception)
            when (exception.Status == StatusCodes.Status404NotFound)
        {
            return null;
        }
    }
}

File: solutions/reference/java/src/main/java/com/microsoft/microhack/catalog/service/AzureBlobImageStore.java
package com.microsoft.microhack.catalog.service;

import com.azure.identity.ManagedIdentityCredentialBuilder;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobServiceClientBuilder;
import com.azure.storage.blob.models.BlobStorageException;
import com.microsoft.microhack.catalog.config.CatalogRuntimeOptions;
import java.io.IOException;
import java.util.Optional;

/** Reads canonical image objects from Blob Storage with the workload identity. */
public final class AzureBlobImageStore implements ImageStore {

    private final BlobContainerClient container;

    public AzureBlobImageStore(CatalogRuntimeOptions options) {
        this(new BlobServiceClientBuilder()
                .endpoint(options.blobServiceEndpoint().toString())
                .credential(new ManagedIdentityCredentialBuilder()
                        .clientId(options.workloadIdentityClientId())
                        .build())
                .buildClient()
                .getBlobContainerClient(options.blobContainerName()));
    }

    AzureBlobImageStore(BlobContainerClient container) {
        this.container = container;
    }

    @Override
    public Optional<byte[]> read(String key) throws IOException {
        if (!LocalImageStore.isCanonicalImageKey(key)) {
            return Optional.empty();
        }
        try {
            return Optional.of(container.getBlobClient(key).downloadContent().toBytes());
        } catch (BlobStorageException exception) {
            if (exception.getStatusCode() == 404) {
                return Optional.empty();
            }
            throw new IOException("Blob image read failed.", exception);
        }
    }
}
